src/parser/ts/index.ts

```diff
@@ -0,0 +1,528 @@
+  );
+};
+
+export function STSReference(query: STSFilter[] = []): STSReference {
+  return {
+    query,
+    type: 'ts',
+  };
+}
+
+export function STSFilter(
+  path: string,
+  value: STSFilter['value'],
+  filter: FilterType = '='
+): STSFilter {
+  return { path, value, filter } as STSFilter;
+}
+
+export function Operator(operator: Operator['operator']): Operator {
+  return { operator };
+}
+
+const getIdFilters = (obj: STSReference) => {
+  return obj.query.filter(isIdsFilter);
+};
+
+const hasIdsFilter = (obj: STSReference) => {
+  return getIdFilters(obj).length;
+};
+
+function isEqualsFilter(query: any): query is STSServerFilter  {
+  return isSTSFilter(query) && query.filter === FilterType.Equals;
+}
+
+function isOneOf(value: string, ...arr: string[]) {
+  return arr.indexOf(value) !== -1;
+}
+
+function isSTSAggregateFilter(query: STSFilter) {
+  return isEqualsFilter(query) && isOneOf(query.path, 'granularity', 'aggregate');
+}
+
+function isIdsFilter(query: STSFilter): query is STSServerFilter {
+  return isEqualsFilter(query) && isOneOf(query.path, 'id', 'externalId');
+}
+
+function isServerFilter(item: STSFilter): item is STSServerFilter {
+  return isEqualsFilter(item) && !isSTSAggregateFilter(item);
+}
+
+function isClientFilter(item: any): item is STSClientFilter {
+  return isSTSFilter(item) && !isEqualsFilter(item);
+}
+
+function isByIdsQuery(query: any): query is STSFilter[] {
+  return isArray(query) && query.some(isIdsFilter);
+}
+
+function isSTSFilter(item: any): item is STSFilter {
+  return isObjectLike(item) && item.path && item.filter && 'value' in item;
+}
+
+function isSTSFilterArr(query: any): query is STSFilter[] {
+  return isArray(query) && query.length && query.every(isSTSFilter);
+}
+
+function isWrappedConst(obj: any): obj is WrappedConst {
+  return isObjectLike(obj) && 'constant' in obj;
+}
+
+function isSTSReference(obj: any): obj is STSReference {
+  return isObjectLike(obj) && obj.type === 'ts';
+}
+
+function isOperator(obj: any): obj is Operator {
+  return isObjectLike(obj) && 'operator' in obj;
+}
+
+function isSTSFunction(obj: any): obj is STSFunction {
+  return isObjectLike(obj) && 'args' in obj && 'func' in obj;
+}
+
+function isSpecificFunction(obj: any, name: string): obj is STSFunction {
+  return isSTSFunction(obj) && name === obj.func;
+}
+
+function isMapFunction(obj: any): obj is MapFunction {
+  return isSpecificFunction(obj, 'map');
+}
+
+function isSumFunction(obj: any): obj is MultiaryFunction {
+  return isSpecificFunction(obj, 'sum');
+}
+
+function isMultiaryFunction(obj: any): obj is MultiaryFunction {
+  return isSTSFunction(obj) && isOneOf(
+    obj.func,
+    'avg',
+    'sum',
+    'min',
+    'max',
+    'pow',
+    'round',
+    'on_error',
+  )
+}
+
+
+type TSResponseMap =  { [s: string]: TimeSeriesResponseItem };
+
+type StringMap = { [key: string]: string };
+
+type STSItem = (STSQueryItem[] | STSQueryItem)[] | STSQueryItem | STSFunction | WrappedConst;
```

Comment:
we don't need `STSFunction` and `WrappedConst` here, right? They're included in `STSQueryItem`

Reply:
oh, thanks. I can actually simplify types even more

Reply:
fixed